lxd/recover: Use configuration defaults

MusicDin opened the pull request:
Extract the defaults generation inside each pool driver's `Create()` function into a separate `PopulateConfig()` function.

On `lxd recover`, the pool's configuration is populated with default values before the configuration gets validated.

Fixes #10862

MusicDin commented:
@tomponline

Please check whether:
- [ ] the approach of extracting default configuration from each pool's driver `Create()` function is followed correctly,
- [ ] `PopulateConfig()` function should contain config validation (e.g. is `zfs.pool_name` valid)

In addition, I have avoided changing `config["source"]` to prevent any changes in the current flow of pool's driver `Create()` function.

tomponline commented on lxd/storage/backend_lxd.go:
```diff
@@ -208,6 +213,12 @@ func (b *lxdBackend) Create(clientType request.ClientType, op *operations.Operat
 	}
 
+	// Populate storage pool configuration with default values
+	err = b.driver.PopulateConfig()
```
this should be done inside the driver I think.

tomponline commented on lxd/storage/drivers/driver_btrfs.go:
```diff
@@ -103,6 +103,11 @@ func (d *btrfs) Info() Info {
 	}
 }
 
+// PopulateConfig populates the storage pool's configuration file with the default values.
```
We would want a separate commit per driver for review clarity.

tomponline commented on lxd/storage/drivers/driver_ceph.go:
```diff
@@ -122,6 +115,17 @@ func (d *ceph) Create() error {
+// Create is called during pool creation and is effectively using an empty driver struct.
+// WARNING: The Create() function cannot rely on any of the struct attributes being set.
+func (d *ceph) Create() error {
+	revert := revert.New()
```
I think we should be calling FillConfig from inside Create, as it already can modify the config.

tomponline commented:
Lets focus on ZFS for now and we can move the logic for generating defaults for the other storage drivers once we've addressed the actual issue we want to with this PR.

tomponline commented:
This is looking good!

tomponline commented:
jenkins: test this please

tomponline commented:
@MusicDin the tests look good for ZFS, do you want to continue extracting out the logic for the other drivers?

MusicDin commented:
Yes, I will go one by one, so that it is easier to review the changes. 

Should I also extract the rest of the defaults for the ZFS pool, such as the `size` and potentially `source`?

tomponline commented:
> Should I also extract the rest of the defaults for the ZFS pool, such as the `size` and potentially `source`?

Yes please.

MusicDin commented:
@tomponline I suppose that validation should not be part of the `FillConfig()` function?

tomponline commented:
> @tomponline I suppose that validation should not be part of the `FillConfig()` function?

Which part were you thinking of?

tomponline commented on lxd/storage/drivers/driver_zfs.go:
```diff
@@ -235,14 +264,6 @@ func (d *zfs) Create() error {
-			d.config["zfs.pool_name"] = d.name
-		}
-
 		// Validate pool_name.
 		if strings.Contains(d.config["zfs.pool_name"], "/") {
 			return fmt.Errorf("zfs.pool_name can't point to a dataset when source isn't set")
```
Ah, no, lets keep validation until after FillConfig has been applied.

MusicDin commented:
@tomponline 

Extracted defaults:
- [x] zfs
- [ ] btrfs
- [ ] dir

If there are no comments, I'll do it for the rest (ceph and lvm)

tomponline commented on lxd/api_internal_recover.go:
```diff
@@ -177,6 +177,12 @@ func internalRecoverScan(s *state.State, userPools []api.StoragePoolsPost, valid
 				}
 
+				// Populate configuration with default values
+				err := pool.Driver().FillConfig()
```
So the commit ordering is a little odd here, as we are calling `FillConfig` before its been added to the storage interface and drivers. Please can you order the commits logically so they dont use resources before they exist.

Also, just a small nit, but can you end comments with a full stop. Thanks

tomponline reviewed (changes requested):
LGTM, just a couple of small nits.

MusicDin commented:
@tomponline 

I have fixed the order of commits and added a full stop at the end of comments.
In addition, I have extracted the defaults for:
- [ ] ceph
- [ ] cephfs
- [ ] cephobject
- [ ] lvm

MusicDin commented:
@tomponline To me the tests look OK. Most failed tests are due to `no space left on device`. Only [this one (TestEndpoints_NetworkSocketBasedActivation)](https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/9929/arch=amd64,backend=dir,compiler=golang-1.18,section=standalone/console) actually failed.

Please take another look, and if there are no other comments, I think we can merge this one?

tomponline commented:
Please can you rebase on master so you get the new github runner system tests.

tomponline reviewed (approved):
LGTM!
